src/Spectre.libWavelet/WaveletDecomposerRef.cpp

```diff
@@ -0,0 +1,126 @@
+/*
+* WaveletDecomposerRef.cpp
+* Reference implementation of wavelet decomposer.
+*
+* Bibliography:
+* https://en.wikipedia.org/wiki/Stationary_wavelet_transform#Implementation
+* http://web.iitd.ac.in/~sumeet/WaveletTutorial.pdf
+* http://www.egr.msu.edu/~aviyente/ece802swt.ppt
+*
+Copyright 2018 Michal Gallus
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+#include "WaveletDecomposerRef.h"
+#include "PrecomputedDaubechiesCoefficients.h"
+
+namespace spectre::algorithm::wavelet
+{
+static constexpr size_t ComputeScale(unsigned level)
+{
+    return 1ULL << (level + 1); // 2^(L + 1)
+}
+
+static constexpr size_t CeiledDivision(size_t dividend, size_t divisor)
```

Comment:
I'm not sure if the `constexpr` really does anything here. There is no invocation involving constants that could be evaluated in the compile time (https://stackoverflow.com/a/14248310) and I doubt that the finite amount of combination of those two arguments is so low that the compiler makes up a look-up table (of course if such scenario is even a thing). The `ComputeScale` function could also be a questionable case, but if compiler goes deep into the semantics, it could find out the `WAVELET_LEVELS` and actually pre-compute those few values.

Reply:
Well, apparently it optimizes it out.
![image](https://user-images.githubusercontent.com/4967505/36062715-53519c0e-0e72-11e8-9183-fced6f82d04b.png)

Reply:
It actually doesn't precompute the value.
![image](https://user-images.githubusercontent.com/4967505/36062728-94cda7fe-0e72-11e8-9785-f54cb63826d9.png)

Reply:
Btw. found some useful bitshifts out there:
![image](https://user-images.githubusercontent.com/4967505/36062961-7d5fe330-0e76-11e8-845e-7fe866ce971a.png)